lib/TPP/FastMathOpts.cpp

```diff
@@ -0,0 +1,136 @@
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+template <typename OP> static bool hasOnlyScalarElementwiseOp(Region &region) {
+  if (!region.hasOneBlock())
+    return false;
+  if (std::distance(region.front().begin(), region.front().end()) != 2)
+    return false;
+  for (Operation &op : region.front()) {
+    if (!isa<OP, linalg::YieldOp>(op) ||
+        llvm::any_of(op.getResultTypes(),
+                     [](Type type) { return !type.isIntOrFloat(); }))
+      return false;
+  }
+  return true;
+}
+
+static bool hasOnlyYield(linalg::GenericOp linalgOp) {
+  Region &region = linalgOp.getRegion();
+  if (!region.hasOneBlock())
+    return false;
+  if (std::distance(region.front().begin(), region.front().end()) != 1)
+    return false;
+  Operation &op = region.front().front();
+  return isa<linalg::YieldOp>(op);
+}
+
+static bool isBroadCastOp(linalg::GenericOp linalgOp) {
+  if (linalgOp->getNumOperands() != 2 || linalgOp->getNumResults() != 1)
+    return false;
+  if (!hasOnlyYield(linalgOp))
+    return false;
+  SmallVector<unsigned> perm;
+  return linalgOp.getMatchingIndexingMap(linalgOp.getDpsInputOperand(0))
+      .isPermutationOfMinorIdentityWithBroadcasting(perm);
+}
+
+static std::optional<linalg::GenericOp> getBroadCastProdcuer(OpOperand *rhs) {
+  linalg::GenericOp broadcastOp = rhs->get().getDefiningOp<linalg::GenericOp>();
+  if (!broadcastOp || !isBroadCastOp(broadcastOp))
+    return std::nullopt;
+  return broadcastOp;
+}
+
+static bool isConstantZero(linalg::FillOp fillOp) {
+  Value cst = fillOp.getInputs()[0];
+  return matchPattern(cst, m_AnyZeroFloat()) || matchPattern(cst, m_Zero());
+}
+
+static bool isFillWithZeros(linalg::Conv2DNhwcHwcfOp convOp) {
+  auto fillOp =
+      convOp.getDpsInitOperand(0)->get().getDefiningOp<linalg::FillOp>();
+  if (!fillOp)
+    return false;
+  return isConstantZero(fillOp);
+}
+
+// Instead of initializing the output of a convolution with zero and then add a
+// bias, initialize the output of the convolution with the bias.
+struct InitializeConv : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (!linalgOp.hasTensorSemantics())
+      return failure();
+    if (linalgOp->getNumOperands() != 3 || linalgOp->getNumResults() != 1)
+      return failure();
+
+    if (!linalg::isElementwise(linalgOp) ||
+        !hasOnlyScalarElementwiseOp<arith::AddFOp>(linalgOp.getRegion()))
+      return failure();
+
+    OpOperand *lhs = linalgOp.getDpsInputOperand(0);
+    OpOperand *rhs = linalgOp.getDpsInputOperand(1);
+
+    auto convProducer = lhs->get().getDefiningOp<linalg::Conv2DNhwcHwcfOp>();
```

Comment:
Is this pattern specific to convolutions? Even, to this particular type of convolution?

Reply:
yes, resnet.